circleguard/visualizer.py

```diff
@@ -24,51 +30,94 @@
 FRAMES_ON_SCREEN = 15  # how many frames for each replay to draw on screen at a time
 PEN_BLACK = QPen(QColor(17, 17, 17))
 PEN_WHITE = QPen(QColor(255, 255, 255))
-X_OFFSET = 64
-Y_OFFSET = 48
-SPEED_OPTIONS = [0.10, 0.25, 0.50, 0.75, 1.00, 1.25, 1.50, 2.00, 5.00, 10.00]
+X_OFFSET = 64+192
+Y_OFFSET = 48+60
+SPEED_OPTIONS = [0.01, 0.10, 0.25, 0.50, 0.75, 1.00, 1.25, 1.50, 2.00, 5.00, 10.00]
+SCREEN_WIDTH = 640+384
+SCREEN_HEIGHT = 480+120
+
+__slider_dir = TemporaryDirectory()
+SLIDER_LIBARY = Library.create_db(get_setting("cache_dir"))
 
 
 class _Renderer(QWidget):
     update_signal = pyqtSignal(int)
 
-    def __init__(self, replays=[], beatmap_path="", parent=None):
+    def __init__(self, replays=[], beatmap_id=None, beatmap_path=None, parent=None):
         super(_Renderer, self).__init__(parent)
-        # initialize variables
-        self.setFixedSize(640, 480)
-        self.replay_amount = len(replays)
-        self.current_time = 0
-        self.pos = [1]*self.replay_amount  # so our first frame is at data[0] since we do pos + 1
-        self.buffer = [[[[0, 0, 0]]]*self.replay_amount][0]
-        self.buffer_additions = [[[[0, 0, 0]]]*self.replay_amount][0]
-        self.clock = clock.Timer()
-        self.last_time = time.time_ns()
+        self.setMinimumSize(SCREEN_WIDTH, SCREEN_HEIGHT)
+        self.painter = QPainter()
+
+        # beatmap init stuff
         self.hitobjs = []
-        self.paused = False
-        self.beatmap_path = beatmap_path
-        self.CURSOR_COLORS = [QPen(QColor().fromHslF(i/self.replay_amount,0.75,0.5)) for i in range(self.replay_amount)]
-        self.playback_len = 0
-        if beatmap_path != "":
-            self.beatmap = Beatmap(beatmap_path)
-            self.playback_len = self.beatmap.hitobjects[-1].time
-        self.data = []
-        self.usernames = []
+        if beatmap_path != None:
+            self.beatmap = Beatmap.from_path(beatmap_path)
+            self.beatmap_flag = True
+            self.playback_len = self.beatmap.hit_objects[-1].time.total_seconds() * 1000 + 3000
+        elif beatmap_id != None:
+            self.beatmap = SLIDER_LIBARY.lookup_by_id(beatmap_id, download=True, save=True)
+            self.beatmap_flag = True
+            self.playback_len = self.beatmap.hit_objects[-1].time.total_seconds() * 1000 + 3000
+        else:
+            self.playback_len = 0
+            self.beatmap_flag = False
+        if not get_setting("render_beatmap"):
+            self.beatmap_flag = False
+        # beatmap stuff
+        if self.beatmap_flag:
```

Comment:
add comment linking to the osu wiki page that has this formula

Reply:
fun fact I can't find the page in the wiki anymore, they removed it for some reason.
https://github.com/ppy/osu-wiki/blob/master/meta/unused/difficulty-settings.md